### PR TITLE
Track average purchase price and add Transfer event handling

### LINKED ISSUE #336: Track average price of each user's staked token position
# Requirement

The subgraph has to be updated to allow computing each use's the unrealized gains on their staked position.

To do so it must track the average price each paid for the share tokens, measured in deposit tokens or assets. The user in this case is the account defined as owner of the shares in the `Deposit` events. Each time the user deposits, the average must be updated proportionally to the current staked amount.

There is no need to track the events related to the exit queue or withdrawals as those will not affect the average purchase price.

## Example

Ratio at t0 is 1 sVUSD = 1 VUSD.
User deposits 100 VUSD and gets 100 sVUSD.
Average price is 1 VUSD/sVUSD.

Ratio at t1 is 1 sVUSD = 1.1 VUSD.
User deposits 110 VUSD and gets 100 sVUSD.
Average price is:

```text
(100 * 1 + 100 * 1.1) / (100 + 100) = 1.05
```

At t2 the user withdraws 50 sVUSD and gets 55 VUSD back
Average price does not change.
New balance is 150 sVUSD.

Ratio at t3 is 1 sVUSD = 1.2 VUSD.
User deposits 120 VUSD and gets 100 sVUSD.
Average price is:

```text
(150 * 1.05 + 100 * 1.2) / (150 + 100) = 1.11
```

Ratio at t4 is 1 sVUSD = 1.5 VUSD.
User deposits 75 VUSD and gets 50 sVUSD.
Average price is:

```text
(250 * 1.11 + 50 * 1.5) / (250 + 50) = 1.175
```

## Calculation

```text
oldShares = currentShares - newShares
depositPrice = assetsIn / newShares
newAveragePrice = (oldShares * oldAveragePrice + newShares * depositPrice) / (oldShares + newShares)
```

In this case at t2:

```text
(150 sVUSD * 1.05 VUSD/sVUSD + 100 sVUSD * 1.2 VUSD/sVUSD) / (150 sVUSD + 100 sVUSD) = 1.11 VUSD/sVUSD
```

## Side note

The earnings can then be computed the difference between the current rate and the average, multiplied by the number of shares:

```text
(1.2 VUSD/sVUSD - 1.11 VUSD/sVUSD) * 250 sVUSD = 22.5 VUSD
```

But the subgraph is primarily a way to index the blockchain and compute values based on its state. For current state, it is easier to query the chain directly. So it will only provide the average purchase price and will let the consumer compute the current earnings by querying on-chain the rest of the data.

## Tasks

- [ ] Update the ABI to include the ERC-4626 Deposit event.
- [ ] Add a new entity to track each user's average purchase price.
- [ ] Add a new handler to process the Deposit event by creating/updating the tracking entity.
- [ ] Add tests. Take example steps above as the basis for the cases.

## Areas of improvement

Computing the new price average from the previous one may suffer from precision loss and drift over time because of division operations and rounding. When storing the average prices with 18 decimals, this problem may be considered negligible.

In order to avoid precision drift a different approach is needed. In this case, both deposits, withdrawal requests and withdrawal cancellations have to be handled to maintain the totals of assets and shares, then derive the price from those totals instead.

## Open questions

Share transfers may affect the calculation. Transfers out are like withdrawals so may not need to be handled. Transfers in would impact the state of each user's average price. In this case, the incoming shares could be valued at zero to mimic what is done in trad-fi.

### PR DESCRIPTION
**Pull request by gabmontes**
### Description

This pull request introduces support for tracking users' average staking price in the subgraph by adding a new entity, event handlers, and comprehensive tests. The main changes include updating the `StakingVault` ABI to include relevant events and functions, implementing handlers for `Deposit` and `Transfer` events to maintain accurate average price data for each user, and extending the schema and tests accordingly.

**Subgraph logic and schema updates:**

* Added a new entity `UserStakingPosition` to the GraphQL schema to track each user's staking vault address, owner, and average price.
* Implemented `handleDeposit` and `handleTransfer` event handlers in `earn.ts` to update the `UserStakingPosition` entity, including logic for calculating and updating the average price per user on deposits and transfers.
* Registered the new entity and event handlers in `subgraph.yaml` to ensure the subgraph indexes `Deposit` and `Transfer` events.

**ABI and data model updates:**

* Updated `StakingVault.json` ABI to include the `balanceOf` function and `Deposit` and `Transfer` events, enabling the subgraph to query balances and react to relevant events. [[1]](diffhunk://#diff-24f9092cd36be19b8e39baa2b63b7d01c7bae3d13e008b2a353f0a67e8830b57R2-R20) [[2]](diffhunk://#diff-24f9092cd36be19b8e39baa2b63b7d01c7bae3d13e008b2a353f0a67e8830b57R53-R83) [[3]](diffhunk://#diff-24f9092cd36be19b8e39baa2b63b7d01c7bae3d13e008b2a353f0a67e8830b57R182-R206)

**Testing and utilities:**

* Added utility functions and comprehensive tests for deposit and transfer scenarios, ensuring correct average price tracking and handling of edge cases (e.g., transfers, mints, withdrawals). [[1]](diffhunk://#diff-f52af6cec58f2945e9daaad57e222833767a088481fc75a16da1d30669e17636R5-R40) [[2]](diffhunk://#diff-f52af6cec58f2945e9daaad57e222833767a088481fc75a16da1d30669e17636R52-R70) [[3]](diffhunk://#diff-b74dd3f95bad2b490bb56c7f921b2d74b20de91c8bbd17d066732522d2f892a2R14-R24) [[4]](diffhunk://#diff-b74dd3f95bad2b490bb56c7f921b2d74b20de91c8bbd17d066732522d2f892a2R54-R63) [[5]](diffhunk://#diff-b74dd3f95bad2b490bb56c7f921b2d74b20de91c8bbd17d066732522d2f892a2R137-R309) [[6]](diffhunk://#diff-b74dd3f95bad2b490bb56c7f921b2d74b20de91c8bbd17d066732522d2f892a2R509-R622)

**Other:**

* Bumped the subgraph package version to 1.3.0.

### Related issue(s)

Closes #336

### Checklist

- [x] Manual testing passed.
- [x] Automated tests added, or N/A.
- [x] Documentation updated, or N/A.
- [x] Environment variables set in CI, or N/A.

### CODE REVIEW

File: subgraph/tests/earn.test.ts [hunk 1]
```diff
@@ -319,3 +506,117 @@ describe("handleWithdrawClaimed", function () {
...
+    handleTransfer(
+      createTransferEvent(
+        zeroAddress,
+        ownerAddress,
+        BigInt.fromString("100000000000000000000"),
+      ),
+    );
+
+    assert.entityCount("UserStakingPosition", 0);
+  });
```
[the remark this answers is not in the document]
Comment (gabmontes) [reply subgraph/tests/earn.test.ts:828]: Cases handled.

File: subgraph/src/earn.ts [hunk 2]
```diff
@@ -52,6 +57,60 @@ export function handleBlock(block: ethereum.Block): void {
...
+  const vault = StakingVault.bind(vaultAddress);
+  const currentShares = vault.balanceOf(owner);
+  const newShares = event.params.shares;
+  const oldShares = currentShares.minus(newShares);
+
+  const calculatedAssets = oldShares
+    .times(entity.averagePrice)
+    .plus(event.params.assets.times(WAD));
+  entity.averagePrice = calculatedAssets.div(currentShares);
+
```
[the remark this answers is not in the document]
Comment (gabmontes) [reply subgraph/src/earn.ts]: This case cannot happen now.

File: subgraph/src/earn.ts [hunk 3]
```diff
@@ -52,6 +57,60 @@ export function handleBlock(block: ethereum.Block): void {
...
+  const vaultAddress = dataSource.address();
+  const to = event.params.to;
+  const id = buildId(vaultAddress, to.toHexString());
+
+  let entity = UserStakingPosition.load(id);
+  if (entity == null) {
+    entity = new UserStakingPosition(id);
+    entity.averagePrice = BigInt.fromI32(0);
+    entity.owner = to;
```
[the remark this answers is not in the document]
Comment (gabmontes) [reply subgraph/src/earn.ts]: No. Handled.

File: subgraph/src/earn.ts [hunk 4]
```diff
@@ -52,6 +57,60 @@ export function handleBlock(block: ethereum.Block): void {
...
+    entity.stakingVaultAddress = vaultAddress;
+  }
+
+  const vault = StakingVault.bind(vaultAddress);
+  const currentShares = vault.balanceOf(to);
+  const oldShares = currentShares.minus(event.params.value);
+
+  // Transferred-in shares are valued at 0, diluting the average price proportionally.
+  entity.averagePrice = entity.averagePrice.times(oldShares).div(currentShares);
```
[the remark this answers is not in the document]
Comment (gabmontes) [reply subgraph/src/earn.ts]: No. Handled.

File: subgraph/src/earn.ts [hunk 5]
```diff
@@ -52,6 +58,68 @@ export function handleBlock(block: ethereum.Block): void {
...
+
```
[the remark this answers is not in the document]
Comment (gabmontes) [reply subgraph/src/earn.ts:137]: Handled!

File: subgraph/src/earn.ts [hunk 6]
```diff
@@ -52,6 +58,72 @@ export function handleBlock(block: ethereum.Block): void {
...
+  let entity = UserStakingPosition.load(id);
+  if (entity == null) {
+    entity = new UserStakingPosition(id);
+    entity.averagePrice = BigInt.fromI32(0);
+    entity.owner = owner;
+    entity.stakingVaultAddress = vaultAddress;
+  }
+
+  const vault = StakingVault.bind(vaultAddress);
+  const currentShares = vault.balanceOf(owner);
```
Comment (gndelia) [inline subgraph/src/earn.ts]: Could this be a misread if a user deposits in the same block that user gets a transfer of shares? Similarly, in the `handleTransfer` if a user gets 2 transfers within the same block. wouldn't `vault.balanceOf(owner)` return the same value in both handlers, skewing the calc of "oldShares" ?

Comment (gabmontes) [reply subgraph/src/earn.ts]: Good catch, Gonza's Claude! Smart boi!

To solve this, the balance will have to be kept in the entity, tracking all deposit/withdraw and transfer events.

Comment (gndelia) [reply subgraph/src/earn.ts]: :agitate:

Comment (gabmontes) [reply subgraph/src/earn.ts]: Fixed!

**Review by ArturDolzan — APPROVED** [review]

File: subgraph/src/earn.ts [hunk 7]
```diff
@@ -52,6 +58,86 @@ export function handleBlock(block: ethereum.Block): void {
...
+  if (event.params.value.equals(BigInt.fromI32(0))) {
+    return;
+  }
+
+  const vaultAddress = dataSource.address();
+  const value = event.params.value;
+
+  // Update the sending party: decrement shares and totalCostBasis proportionally.
+  const fromId = buildId(vaultAddress, event.params.from.toHexString());
+  const fromEntity = UserStakingPosition.load(fromId)!;
```
Comment (gndelia) [inline subgraph/src/earn.ts:102]: I did not know `!` was also part of assembly script!

**Review by gndelia — APPROVED** [review]
